cmake/InstallDirs.cmake

```diff
@@ -0,0 +1,52 @@
+if(WIN32)
+  if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
+    set(program_files "$ENV{ProgramFiles}")
+
+    # For 32 bit builds.
+    if(CMAKE_SIZEOF_VOID_P EQUAL 4 AND ENV{ProgramFiles\(x86\)})
+      set(program_files "$ENV{ProgramFiles\(x86\)}")
+    endif()
+
+    if(NOT program_files)
+      if(NOT CMAKE_SIZEOF_VOID_P EQUAL 4)
+        set(program_files "/Program Files")
+      else()
+        set(program_files "/Program Files (x86)")
+      endif()
+    endif()
+
+    file(TO_CMAKE_PATH "${program_files}/ccache" CMAKE_INSTALL_PREFIX)
+
+    set(CMAKE_INSTALL_PREFIX "${CMAKE_INSTALL_PREFIX}" CACHE STRING "System-wide installation prefix" FORCE)
+  endif()
+
+  if(NOT CMAKE_INSTALL_SYSCONFDIR)
+    set(program_data "$ENV{ALLUSERSPROFILE}")
+
+    if(NOT program_data)
+      set(program_data "/ProgramData")
+    endif()
+
+    file(TO_CMAKE_PATH "${program_data}/ccache" CMAKE_INSTALL_SYSCONFDIR)
+
+    set(CMAKE_INSTALL_SYSCONFDIR "${CMAKE_INSTALL_SYSCONFDIR}" CACHE PATH "System-wide config file location" FORCE)
+  endif()
+
+  set(CMAKE_INSTALL_BINDIR     "" CACHE PATH "executables subdirectory" FORCE)
+  set(CMAKE_INSTALL_SBINDIR    "" CACHE PATH "system administration executables subdirectory" FORCE)
+  set(CMAKE_INSTALL_LIBEXECDIR "" CACHE PATH "dependent executables subdirectory" FORCE)
+  set(CMAKE_INSTALL_LIBDIR     "" CACHE PATH "object libraries subdirectory" FORCE)
+endif()
+
+include(GNUInstallDirs)
+
+# Escape SYSCONFDIR for a C macro in config.h .
```

Comment:
Why is the escaping needed? You can have the `CMAKE_INSTALL_SYSCONFDIR` which uses slashes and the `std::filesystem::path` will do the right thing. 

In my opinion this adds code that is not needed an complicates things.

Reply:
Just tried this, it actually doesn't:

```cpp
#include <iostream>
#include <filesystem>

int main(int argc, char** argv)
{
    std::cout << (std::filesystem::path{"C:/ProgramData"} / "ccache").string() << std::endl;
    return 0;
}
```

, the result is:

```console
C:/ProgramData\ccache
```

. Also, if it did work it would require always passing the path through something before using it, I'd rather have it be in the best form in the first place.

I could probably define the macro as a `constexpr` function that transforms the slashes, but that just moves the complexity from cmake to C++. But I can do that too.

Reply:
Also tried:

```cpp
std::filesystem::path{"C:/ProgramData", std::filesystem::path::native_format}
```

, but that also doesn't convert the slash.

Reply:
ok, then `std::filesystem::path` is sillier than I thought.

This whole exercise is purely academic since `ALLUSERSPROFILE` will always exist and this value will never be used.

Reply:
This works though:

```cpp
#include <iostream>
#include <filesystem>

int main(int argc, char** argv)
{
    std::cout << (std::filesystem::path{"C:/ProgramData"} / "ccache")
        .lexically_normal().string() << std::endl;
    return 0;
}
```

, will add that to the wrapper.

Thanks for the idea!

Reply:
I can remove this now if you really want as the macro is only used in one place, and I made it go through the normalizing `PATH(...)` wrapper.

Reply:
I really think the value should be correct at build time. Another thing I could do is hide the gory details by making a `configure_file()` wrapper that does this automatically. @jrosdahl what do you suggest?

Reply:
@cristianadam I removed the unnecessary cache entry and put the two lines into the `GenerateConfigurationFile.cmake` file where the `configure_file()` call is, is this more tolerable?

Reply:
Do I understand correctly that you want the path printed in logs or by `ccache -sv` to look nice (be in backslash form) in case `%ALLUSERSPROFILE%` isn't set? (I assume that using the unnormalized path works fine.) I agree with @cristianadam that it doesn't sound very important, but I'm fine with keeping the two lines if you like.

Reply:
Another option would perhaps be to let the CMake script require that `ALLUSERSPROFILE` is set and exit with an error otherwise?

Reply:
`%ALLUSERSPROFILE%` will always be set with nearly 100% probability. My only issue was that the path written to `config.h` be in the correct form. If you're fine with the two lines I'll keep them, we will need to revisit all of this again anyway, see below.

Reply:
OK.